compiler/qsc_parse/src/lex/formatter.rs

```diff
@@ -0,0 +1,270 @@
+        }
+    } else {
+        indent_level
+    };
+
+    match (&left.kind, &right.kind) {
+        (
+            ConcreteTokenKind::Cooked(TokenKind::Open(l)),
+            ConcreteTokenKind::Cooked(TokenKind::Close(r)),
+        ) if l == r => {
+            rule_no_space(left, whitespace, right, &mut edits);
+        }
+        (ConcreteTokenKind::Comment | ConcreteTokenKind::Cooked(TokenKind::DocComment), _) => {
+            rule_trim_comments(left, &mut edits, code);
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (ConcreteTokenKind::Cooked(TokenKind::Semi), _) => match &right.kind {
+            ConcreteTokenKind::Comment => {
+                if whitespace.contains('\n') {
+                    rule_indentation(left, whitespace, right, &mut edits, indent_level);
+                }
+            }
+            _ => {
+                rule_indentation(left, whitespace, right, &mut edits, indent_level);
+            }
+        },
+        (_, ConcreteTokenKind::Cooked(TokenKind::Close(Delim::Brace))) => {
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (
+            _,
+            ConcreteTokenKind::Cooked(TokenKind::Keyword(
+                Keyword::Operation | Keyword::Function | Keyword::Newtype | Keyword::Namespace,
+            )),
+        ) => {
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (ConcreteTokenKind::Cooked(TokenKind::Open(Delim::Brace)), _) => {
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (
+            ConcreteTokenKind::Cooked(TokenKind::Close(Delim::Brace)),
+            ConcreteTokenKind::Cooked(TokenKind::Semi),
+        ) => {
+            rule_no_space(left, whitespace, right, &mut edits);
+        }
+        (ConcreteTokenKind::Cooked(TokenKind::Close(Delim::Brace)), _) => {
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (ConcreteTokenKind::Cooked(cooked_left), ConcreteTokenKind::Cooked(cooked_right)) => {
+            match (cooked_left, cooked_right) {
+                (TokenKind::Ident, TokenKind::Ident)
+                | (TokenKind::Keyword(_), TokenKind::Ident)
+                | (_, TokenKind::Colon)
+                | (TokenKind::Colon, _)
+                | (TokenKind::Comma, _) => {
+                    rule_single_space(left, whitespace, right, &mut edits);
+                }
+                (TokenKind::Ident, TokenKind::Open(Delim::Paren))
+                | (TokenKind::Ident, TokenKind::Open(Delim::Bracket))
+                | (TokenKind::Ident, TokenKind::Comma)
+                | (TokenKind::Open(_), _)
+                | (_, TokenKind::Close(_)) => {
+                    rule_no_space(left, whitespace, right, &mut edits);
+                }
+                _ => {}
+            }
+        }
+        _ => {}
```

Comment:
You'd still want to fix the indentation for the catch-all cases right?

Reply:
In all other cases, we don't want to be authoritative about whitespace. The `rule_indentation` removes any whitespace other than new-lines and indentation, so it wouldn't work well to have that be called in situations where we don't want that behavior, like in `let x = 3;`. Those spaces are desired in that example.